nengo_spinnaker/utils/passthrough_nodes.py

```diff
@@ -0,0 +1,154 @@
+
+        # If the target is a passthrough Node then add a connection
+        if isinstance(obj, nengo.Node) and obj.output is None:
+            with network:
+                nengo.Connection(probe.target, probe, seed=probe.seed,
+                                 synapse=probe.synapse)
+
+
+def combine_lti_synapses(a, b):
+    """Combine two LTI filters."""
+    # Assert that both the synapses are LTI synapses
+    assert (isinstance(a, nengo.synapses.LinearFilter) and
+            isinstance(b, nengo.synapses.LinearFilter))
+
+    # Combine
+    return nengo.synapses.LinearFilter(polymul(a.num, b.num),
+                                       polymul(a.den, b.den))
+
+
+def remove_passthrough_nodes(network):
+    """Return a new network with all the passthrough Nodes removed and a
+    mapping from new connections to connections the decoders of which they can
+    use.
+    """
+    # Get a list of the connections and a complete map of objects to inputs and
+    # outputs.
+    conns = list(network.all_connections)
+    inputs, outputs = find_all_io(conns)
+
+    # Prepare to create a map of which connections can just use the decoders
+    # from an earlier connection.
+    connection_decoders = dict()
+
+    # Create a new (flattened) network containing all elements from the
+    # original network apart from passthrough Nodes.
+    with nengo.Network() as m:
+        # Add all of the original ensembles
+        for ens in network.all_ensembles:
+            m.add(ens)
+
+        # Add all of the original probes
+        for probe in network.all_probes:
+            m.add(probe)
+
+    # For all the Nodes, if the Node is not a passthrough Node we add it as
+    # usual - otherwise we combine remove it and multiply together its
+    # input and output connections.
+    for node in network.all_nodes:
+        if node.output is not None:
+            with m:
+                m.add(node)
+            continue
+
+        # Remove the original connections associated with this passthrough
+        # Node from both the list of connections but also the lists
+        # associated with their pre- and post- objects.
+        conns_in = list(inputs[node])
+        conns_out = list(outputs[node])
+
+        for c in conns_in:
+            conns.remove(c)
+            outputs[c.pre_obj].remove(c)
+
+        for c in conns_out:
+            conns.remove(c)
+            inputs[c.post_obj].remove(c)
+
+        # For every outgoing connection
+        for out_conn in outputs[node]:
+            # For every incoming connection
+            for in_conn in inputs[node]:
+                # Create a new transform for the combined connections.  If
+                # the transform is zero then we don't bother adding a new
+                # connection and instead move onto the next combination.
+                transform = np.dot(
+                    full_transform(out_conn),
+                    full_transform(in_conn, slice_pre=False)
```

Comment:
@tcstewar - should be `slice_pre=is_passthrough_node(in_conn.pre_obj)`?

Valid cases where `slice_pre` should be False are:
```python
a = nengo.Ensemble(100, 4)
b = nengo.Node(size_in=1)
nengo.Connection(a[1:3], b, function=lambda x: x[0] * x[1])
```